### ISSUE TITLE
datetime.timestamp() doesn't have enough precision to represent datetime.max

### ISSUE THREAD

**Issue by cc0abac9-1ca2-444c-b861-a6cf55d5d9cf** [python/cpython#91012]
BPO | [46856](https://bugs.python.org/issue46856)
--- | :---
Nosy | @ericvsmith

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-02-25.16:10:01.560>
labels = ['type-bug', 'library', '3.9']
title = 'datetime.max conversion'
updated_at = <Date 2022-03-01.15:47:18.405>
user = 'https://bugs.python.org/jorisgeysens'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-01.15:47:18.405>
actor = 'joris.geysens'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2022-02-25.16:10:01.560>
creator = 'joris.geysens'
dependencies = []
files = []
hgrepos = []
issue_num = 46856
keywords = []
message_count = 6.0
messages = ['414013', '414023', '414032', '414043', '414044', '414275']
nosy_count = 2.0
nosy_names = ['eric.smith', 'joris.geysens']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue46856'
versions = ['Python 3.9']
```

</p></details>

**Comment by psarka** [comment]
The min range of the datetime is also affected:
```
>>> import datetime as dt
>>> dt.datetime.min
datetime.datetime(1, 1, 1, 0, 0)
>>> dt.datetime.min.timestamp()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: year 0 is out of range

```

**Comment by donBarbos** [comment]
I think it's important to note that the problem is not with the `fromtimestamp` method, the point is that `datetime.timestamp` returns a rounded timestamp because 253402300800 is exactly 10000 years, but we asked timestamp for 9999 year 12, 31, ....

**Comment by donBarbos** [comment]
we have a method `timedelta.total_seconds` (we using this method in `datetime.timestamp` like this `return (self - _EPOCH).total_seconds()`):
```python
    def total_seconds(self):
        """Total seconds in the duration."""
        return ((self.days * 86400 + self.seconds) * 10**6 + self.microseconds) / 10**6
```
it returns the rounded number of seconds, the question arose: should we do this? (maybe i should open new issue)

in my opinion rounding up is the wrong approach in the module responsible for date/time

**Comment by StanFromIreland** [comment]
Duplicate of: #89510, https://github.com/python/cpython/issues/88994 ?

**Comment by StanFromIreland** [comment]
We are hitting the limit of Python's float anyway, but for microseconds below that we could just append them?

```python
>>> float("-30610222479.99999")
-30610222479.99999
>>> float("-30610222479.999999")
-30610222480.0
```

What if we dropped a digit off of `max`, that would suffice?

**Comment by picnixz** [comment]
There is a small distinction between `shell` (https://pygments.org/docs/lexers/#pygments.lexers.shell.BashLexer) for shell scripts and `shell-session` for console (https://pygments.org/docs/lexers/#pygments.lexers.shell.BashSessionLexer)

**Comment by picnixz** [comment]
Using https://pygments.org/demo/: With `Bash` (equivalent to `shell`):

![Image](https://github.com/user-attachments/assets/a0865f5a-c098-437a-bfdb-dd8c58b18695)


With `Bash Session` (equivalent to `shell-session`):

![Image](https://github.com/user-attachments/assets/bbe56058-e8e4-44d1-8721-e41623d816e8)

**Comment by picnixz** [comment]
To avoid this weird formatting, I would advise using the full name `zypper install` instead of `in` (though usually we only write `in` indeed)

**Comment by picnixz** [comment]
Oups wrong issue

**Comment by StanFromIreland** [comment]
@picnixz Confused me for a second lol. You should hide them as off-topic and post the in the right place :-)

**Comment by ericvsmith** [comment]
> What if we dropped a digit off of `max`, that would suffice? We currently have a `0 <= microsecond < 1000000` bound on microseconds what if we make it 0-99999?

I don't think we want to do that: it unnecessarily complicates the code and might have unintended effects. I don't think there's a practical problem here that needs addressing.

**Comment by StanFromIreland** [comment]
The problem still exists, if a user specifies too many microseconds that float cannot handle --  it rounds into the next day.

**Comment by picnixz** [comment]
> The problem still exists, if a user specifies too many microseconds that float cannot handle -- it rounds into the next day.

If too many microseconds are specified, and rounding is not wanted, that's an API misuse IMO. The same could be said for floats themselves. If you specify too many digits, then the result is rounded.

> What if we dropped a digit off of max, that would suffice? We currently have a 0 <= microsecond < 1000000 bound on microseconds what if we make it 0-99999?

Note that this would exclude `float("-30610222479.999998")` which would still be valid as it's `-30610222479.99999618530273437500`. So, if we were to change the limit, it should be `< 999999` instead of `1000000`. But again, I don't think we need to overcomplicate the code for this specific event (I don't know when, in practice, we would have this issue and how frequent it would appear).

**Comment by StanFromIreland** [comment]
Do we just want to note this in the docs and close?

**Comment by ericvsmith** [comment]
I personally don't think it's worth a note in the docs. It's how floating point works. How many places would we make this notation?

**Comment by StanFromIreland** [comment]
Maybe just once in the notes at the start "due to the limitations of double and Python's float microseconds may be rounded" though I still think there should be a better approach stopping the rounding.

**Comment by encukou** [comment]
In the `time` module, the solution was to add `time_ns`, and other `*_ns` variants for all functions that dealt with `float` timestamps. 
But `time` deals with *current* times, and the [limits of the internal underlying C type](https://docs.python.org/3/c-api/time.html#c.PyTime_t) are not enough for `datetime`, so any `*_ns` functions would need to be implemented differently. I don't think that's worth it.

Ultimately, `datetime.timestamp()` returns `float`, and `float` is subject to rounding. If you don't want a lossy export, don't use `timestamp()`.
